packages/dart_frog_cli/lib/src/commands/new/new.dart

```diff
@@ -0,0 +1,157 @@
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:args/command_runner.dart';
+import 'package:dart_frog_cli/src/command.dart';
+import 'package:dart_frog_cli/src/commands/commands.dart';
+import 'package:dart_frog_cli/src/commands/new/templates/dart_frog_new_bundle.dart';
+import 'package:mason/mason.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+
+/// {@template new_command}
+/// `dart_frog new <route|middleware> "path/to/route"`
+///
+/// Creates a new route or middleware for dart_frog.
+/// {@endtemplate}
+class NewCommand extends DartFrogCommand {
+  /// {@macro new_command}
+  NewCommand({
+    super.logger,
+    GeneratorBuilder? generator,
+  }) : _generator = generator ?? MasonGenerator.fromBundle {
+    addSubcommand(newRouteCommand = _NewSubCommand('route'));
+    addSubcommand(newMiddlewareCommand = _NewSubCommand('middleware'));
+  }
+
+  final GeneratorBuilder _generator;
+
+  @override
+  String get description => 'Create a new route or middleware for dart_frog';
+
+  @override
+  String get name => 'new';
+
+  @override
+  final String invocation = 'dart_frog new <route|middleware> "path/to/route"';
+
+  /// Subcommand for creating a new route.
+  @visibleForTesting
+  late final DartFrogCommand newRouteCommand;
+
+  /// Subcommand for creating a new middleware.
+  @visibleForTesting
+  late final DartFrogCommand newMiddlewareCommand;
+}
+
+class _NewSubCommand extends DartFrogCommand {
+  _NewSubCommand(this.name);
+
+  @override
+  String get description => 'Create a new $name for dart_frog';
+
+  @override
+  final String name;
+
+  @override
+  NewCommand get parent => super.parent! as NewCommand;
+
+  @override
+  late final String invocation = 'dart_frog new $name "path/to/route"';
+
+  @visibleForTesting
+  @override
+  // ignore: invalid_use_of_visible_for_testing_member
+  ArgResults? get testArgResults => parent.testArgResults;
+
+  @override
+  // ignore: invalid_use_of_visible_for_testing_member
+  String? get testUsage => parent.testUsage;
+
+  String get _routePath {
+    final rest = results.rest;
+    if (rest.isEmpty) {
+      throw UsageException(
+        'Provide a route path for the new $name',
+        usageString,
+      );
+    }
+    final routeName = rest.first;
+    if (routeName.isEmpty) {
+      throw UsageException(
+        'Route path must not be empty',
+        usageString,
+      );
+    }
+
+    if (!routeName.isValidRoutePath()) {
+      throw UsageException('Route path must be a valid path segment.', '');
```

Comment:
I think it is a nice convention to include a solution suggestion in the error message. Similar to how the other Usage Exception regarding a missing route directory provides a hint (i.e "Make sure to run this command on a dart_frog project.")

What do you think of something along the lines of:

```dart
      throw UsageException('''Route path must be a valid path segment. The path can not be empty, contain a dollar sign or bla and foo''','');
```

Alternatively, instead of combining all of them another option could be to throw a UsageException for each individual check.